### PR TITLE
Unified `ToolMemory` and dynamic artifact storage

### PR DESCRIPTION
**Pull request by vasinov**
## Important PR Changes

* Removed separate memory for text and blob artifacts.
* Unified all tool memory under a single `ToolMemory` abstraction. This change introduces `ToolMemory.artifact_storage`, which is a `dict` with keys for artifact types and values for `BaseArtifactStorage`. The user can now define custom storage backends for any type of tool output.
* Added separate storage abstractions for blob and text artifacts. The idea here is that every storage must implement a set of methods (such as `can_store`, `store_artifact`, `load_artifacts`, `summarize`, and `query`) that can be applied to artifacts that that storage supports. For example, querying text storage will use a text summarizer, while a hypothetical `VideoArtifactStorage` will have a completely different way of summarizing hypothetical `VideoArtifact`s.
* Added empty output handling for tools: memory now returns `InfoArtifact`s back to the LLM explicitly stating that the result was empty.
* Removed extraction activities from `ToolMemoryActivitiesMixin`.

## Future Direction

This PR enables us to start working on a number of important updates over the next month:

* Remove memory activities and refactor them into purpose-built tools for summarization, querying, and extraction. I started to refer to those are "meta tools."
* Add a way to automatically add meta tools as defaults. Perhaps, every non-meta tool should have a list of default meta-tools.
* Elevate tool input/output layer into an OpenAPI interface that every tool will be able to convert to.
* Provide a way to run tool APIs as separate processes/services.
* Implement summarization and querying methods for `BlobArtifactStorage` based on the metadata associated with stored blob artifacts.

Here is a separate list o minor followup changes that we should address on:

* Add better summarization and querying capabilities to `BlobArtifactStorage`.
* Add tool output artifact metadata to memory.
* Refactor `uses_default_memory` out.

---
This PR closes https://github.com/griptape-ai/griptape/issues/315, closes https://github.com/griptape-ai/griptape/issues/294, closes https://github.com/griptape-ai/griptape/issues/360, and closes https://github.com/griptape-ai/griptape/issues/283

### CODE REVIEW

File: griptape/memory/tool/tool_memory.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+@define
+class ToolMemory(ToolMemoryActivitiesMixin, ActivityMixin):
+    name: str = field(
+        default=Factory(lambda self: self.__class__.__name__, takes_self=True),
+        kw_only=True,
+    )
+    artifact_storage: dict[Type, BaseArtifactStorage] = field(factory=dict, kw_only=True)
+    namespace_storage: dict[str, BaseArtifactStorage] = field(factory=dict, kw_only=True)
+    namespace_metadata: dict[str, any] = field(factory=dict, kw_only=True)
```
Comment (collindutter) [inline griptape/memory/tool/tool_memory.py:22]: Use `typing.Any`. See https://github.com/griptape-ai/griptape/issues/360

File: griptape/memory/tool/tool_memory.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+    @artifact_storage.validator
+    def validate_artifact_storage(self, _, artifact_storage: dict[Type, BaseArtifactStorage]) -> None:
+        seen_types = []
+
+        for storage in artifact_storage.values():
+            if type(storage) in seen_types:
+                raise ValueError("Can't have more than memory storage of the same type")
```
Comment (collindutter) [inline griptape/memory/tool/tool_memory.py:30]: Lowercase error

File: griptape/memory/tool/tool_memory.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        else:
+            return find_storage(artifact)
+
+    def process_output(
+            self,
+            tool_activity: callable,
+            subtask: ActionSubtask,
+            output_artifact: BaseArtifact
+    ) -> BaseArtifact:
+        from griptape.utils import J2
```
Comment (collindutter) [inline griptape/memory/tool/tool_memory.py:54]: What's the reason for importing here?

File: griptape/memory/tool/tool_memory.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+        tool_name = tool_activity.__self__.name
+        activity_name = tool_activity.name
+        namespace = output_artifact.name
+
+        if output_artifact.is_empty():
+            return InfoArtifact("Tool output is empty.")
```
Comment (collindutter) [inline griptape/memory/tool/tool_memory.py:61]: Lowercase info string

File: griptape/memory/tool/tool_memory.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+                tool_name=tool_name,
+                activity_name=activity_name,
+                artifact_namespace=namespace
+            )
+
+            return InfoArtifact(output)
+        else:
+            logging.info(f"Output of {tool_name}.{activity_name} can't be stored in {self.name}")
+
+            return ErrorArtifact("Error processing tool output.")
```
Comment (collindutter) [inline griptape/memory/tool/tool_memory.py:76]: Lowercase string.

File: griptape/memory/tool/tool_memory.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+        storage = self.get_storage_for(artifact)
+
+        if namespace_storage and namespace_storage != storage:
+            logging.warning(f"Incompatible storage types")
+
+            return False
+        else:
+            if storage:
+                if isinstance(artifact, ListArtifact):
+                    [
```
Comment (collindutter) [inline griptape/memory/tool/tool_memory.py:89]: I know we do this in other areas of the codebase, but I've got mixed feelings on list comprehension without assignment. I think I'm leaning towards encouraging regular for loops.

File: griptape/tools/base_tool.py [hunk 7]
```diff
@@ -71,13 +70,18 @@ def abs_file_path(self):
     def abs_dir_path(self):
         return os.path.dirname(self.abs_file_path)
 
-    def before_execute(self, activity: callable, value: Optional[dict]) -> Optional[dict]:
-        return value
-
     def execute(self, activity: callable, subtask: ActionSubtask) -> BaseArtifact:
```
Comment (collindutter) [inline griptape/tools/base_tool.py:73]: If we're renaming `before/after_execute` to `before/after_run`, we should rename `execute` to `run`

File: griptape/memory/tool/storage/blob_artifact_storage.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
+from attr import define, field
+from griptape.artifacts import BaseArtifact, ListArtifact, BlobArtifact, InfoArtifact
+from griptape.memory.tool.storage import BaseArtifactStorage
+
+
+@define
+class BlobArtifactStorage(BaseArtifactStorage):
```
Comment (collindutter) [inline griptape/memory/tool/storage/blob_artifact_storage.py:7]: Should this be named `LocalBlobArtifactStorage`? Or will the storage destinations come in the form of Drivers in the future? i.e. `LocalBlobArtifactStorageDriver`, `S3BlobArtifactStorageDriver`

File: griptape/artifacts/base_artifact.py [hunk 9]
```diff
@@ -57,6 +57,9 @@ def from_dict(cls, artifact_dict: dict) -> BaseArtifact:
     def from_json(cls, artifact_str: str) -> BaseArtifact:
         return cls.from_dict(json.loads(artifact_str))
 
+    def is_empty(self) -> bool:
```
[the remark this answers is not in the document]
Comment (vasinov) [reply griptape/artifacts/base_artifact.py:60]: Yes, this is a much better approach.

File: griptape/artifacts/csv_row_artifact.py [hunk 10]
```diff
@@ -30,3 +30,6 @@ def to_dict(self) -> dict:
         from griptape.schemas import CsvRowArtifactSchema
 
         return dict(CsvRowArtifactSchema().dump(self))
+
+    def is_empty(self) -> bool:
```
[the remark this answers is not in the document]
Comment (vasinov) [reply griptape/artifacts/csv_row_artifact.py:34]: I'll add it as a separate issue.

File: griptape/memory/tool/tool_memory.py (hunk 3 above)
Comment (vasinov) [reply griptape/memory/tool/tool_memory.py:54]: Circular import dependency error.

File: griptape/memory/tool/tool_memory.py (hunk 6 above)
Comment (vasinov) [reply griptape/memory/tool/tool_memory.py:89]: I agree.

File: griptape/tools/base_tool.py (hunk 7 above)
Comment (vasinov) [reply griptape/tools/base_tool.py:73]: We already have `run`. `Execute` runs a before hook, `run`, and an after hook. We also use the same pattern in tasks.

File: griptape/memory/tool/storage/blob_artifact_storage.py (hunk 8 above)
Comment (vasinov) [reply griptape/memory/tool/storage/blob_artifact_storage.py:7]: Yeah, I'd like to keep storages high level and then add driver support for individual storage backends later.

**Review by collindutter — APPROVED** [review]

**Pull request merged**